lib/astc_encode.cpp

```diff
@@ -0,0 +1,963 @@
+        uint32_t height = MAX(1, This->baseHeight >> level);
+        uint32_t depth = MAX(1, This->baseDepth >> level);
+        uint32_t faceSlices = This->numFaces == 1 ? depth : This->numFaces;
+
+        priv._levelIndex[level].byteLength             = 0;
+        priv._levelIndex[level].uncompressedByteLength = 0;
+
+        ktx_size_t size = astcBufferSize(width, height, depth,
+                                         block_size_x, block_size_y, block_size_z);
+
+        for (uint32_t layer = 0; layer < MAX(1, This->numLayers); layer++) {
+            for (ktx_uint32_t slice = 0; slice < faceSlices; slice++) {
+                ktx_size_t offset;
+                ktxTexture2_GetImageOffset((ktxTexture2*)This, level, layer,
+                                           slice, &offset);
+
+                // TODO: Fix depth for 3D textures
+                // Create compressor readable image from each image in the container
+                astcenc_image *input_image = nullptr;
+                if (num_components == 1)
+                    input_image = unorm8x1ArrayToImage(This->pData + offset,
+                                                       width, height);
+                else if (num_components == 2)
+                    input_image = unorm8x2ArrayToImage(This->pData + offset,
+                                                       width, height);
+                else if (num_components == 3)
+                    input_image = unorm8x3ArrayToImage(This->pData + offset,
+                                                       width, height);
+                else // assume (num_components == 4)
+                    input_image = unorm8x4ArrayToImage(This->pData + offset,
+                                                       width, height);
+
+                assert(input_image);
+
+                astc_input_images.push_back(input_image);
+                astc_image_sizes.push_back(size);
+
+                astc_images_size += size;
+
+                priv._levelIndex[level].byteLength             += size;
+                priv._levelIndex[level].uncompressedByteLength += size;
+            }
+        }
+
+        // Offset is from start of the astc compressed block
+        priv._levelIndex[level].byteOffset = level_offset;
+        priv._requiredLevelAlignment = 8; // For astc its always 8 bytes (128bits) irrespective of block sizes
+
+        ktx_size_t prev_level_offset = level_offset;
+        level_offset += _KTX_PADN(priv._requiredLevelAlignment,
+                                  priv._levelIndex[level].byteLength);
+
+        ktx_size_t diff = level_offset - prev_level_offset;
+
+        assert(diff - priv._levelIndex[level].byteLength == 0
+               && "\nSome format/blocks aren't alighned to 8 bytes, need to create offsets\n");
+    }
+
+    free(This->pData); // No longer needed. Reduce memory footprint.
+    This->pData = NULL;
+    This->dataSize = 0;
+
```

Comment:
This is where you should create a prototype texture of the correct VkFormat. Doing so will create a texture with the correct DFD, properly sized data allocation and the level index already set up. Plus you can use the level offset, etc functions on the prototype to see where things need to go in the data. See ktxTexture2_TranscodeBasis to see what to do. It simplifies things.

Reply:
Fixed in https://github.com/KhronosGroup/KTX-Software/pull/433/commits/d56a6e9465b0875bdc9f518ca269d2e5b1f2d401

Reply:
The Files Changed view doesn't seem to be showing me any changes with prototype in them.

Reply:
Are you looking at the right commit? It should be in there. https://github.com/KhronosGroup/KTX-Software/commit/d56a6e9465b0875bdc9f518ca269d2e5b1f2d401 commit introduces `prototype` on line 705